pkg/controller/admissionchecks/multikueue/workload.go

```diff
@@ -279,11 +293,43 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
 			Reason:  remoteFinishedCond.Reason,
 			Message: fmt.Sprintf("From remote %q: %s", remote, remoteFinishedCond.Message),
 		})
-		return a.client.Status().Patch(ctx, wlPatch, client.Apply, client.FieldOwner(ControllerName+"-finish"), client.ForceOwnership)
+		return reconcile.Result{}, a.client.Status().Patch(ctx, wlPatch, client.Apply, client.FieldOwner(ControllerName+"-finish"), client.ForceOwnership)
 	}
 
 	hasReserving, reservingRemote := group.FirstReserving()
 
+	// if the reserving workload is out of sync, remove it
+	if hasReserving {
+		outOfSync := group.local == nil || !equality.Semantic.DeepEqual(group.local.Spec, group.remotes[reservingRemote].Spec)
```

Comment:
Will this not be done by GC?

Reply:
Isn't this part `group.local == nil` covered by the MK GC? IIRC MK GC removes workloads on workers which don't have the local counterpart, so seems to be the case here.

Reply:
To avoid nil pointer we can have `group.local != nil && !equality.Semantic.DeepEqual(group.local.Spec, group.remotes[reservingRemote].Spec)`, so it is not convincing argument.

> I cannot understand why we wold like to wait fro the periodic GC to run when we can do the cleanup now.

Sorry, for now I just asked the question to understand what is happening (not claiming which one is better in the comments above) because I wasn't sure of my understanding. Can you summarize the pros & cons for deleting now and leaving this to MKGC?

Reply:
Can you?

Reply:
> If left, the workload will remain queued in the worker cluster and can potentially block quota

Right, this is true in the other cases as well, I'm wondering what makes this situation different. What is the scenario leading to "group.local == nil"?

> We cannot consider a non-nil wl being Semantic.DeepEqual to a nil.

I don't think I suggested that (at least not intentionally). I'm saying (1) nil local workload -> MKGC, (2)  !Semantic.DeepEqual -> Delete remote. It is actually the current code that interprets (1) and (2) as the same (outOfSync). Unless I'm missing something.

Reply:
> Right, this is true in the other cases as well, I'm wondering what makes this situation different.

We can do something to avoid it. 

The function was refactored, please check it.

Reply:
Oh, the code with the check is gone, feel free to resolve.